File: inference_profiler/collectors/proc.py
import glob
import json
import os

from .base import BaseCollector


class ProcCollector(BaseCollector):
    @staticmethod
    def collect():
        process_list = []
        process_count = 0

        for pid_path in glob.glob('/proc/[0-9]*'):
            try:
                pid = int(os.path.basename(pid_path))
                stat_data, t_stat_data = BaseCollector._probe_file(os.path.join(pid_path, 'stat'))
                cmdline, t_cmdline = BaseCollector._probe_file(os.path.join(pid_path, 'cmdline'))
                status_data, t_status_data = BaseCollector._parse_proc_kv(os.path.join(pid_path, 'status'))
                statm_data, t_statm = BaseCollector._probe_file(os.path.join(pid_path, 'statm'))

                if not stat_data:
                    continue

                rparen_index = stat_data.rfind(')')

                if rparen_index == -1:
                    continue

                stat_parts = stat_data[rparen_index + 2:].split()

                if len(stat_parts) < 40:
                    continue

                process_count += 1

                # Parse statm for resident set size
                # statm fields: size resident shared text lib data dt (all in pages)
                rss_pages = 0
                if statm_data:
                    statm_parts = statm_data.split()
                    if len(statm_parts) >= 2:
                        try:
                            rss_pages = int(statm_parts[1])
                        except ValueError:
                            pass

                # Get page size (usually 4096)
                page_size = os.sysconf('SC_PAGE_SIZE') if hasattr(os, 'sysconf') else 4096
                rss_bytes = rss_pages * page_size

                process_list.append({
                    # Process identification
                    "pId": pid,
                    "pName": status_data.get('Name', stat_data[stat_data.find('(') + 1:rparen_index]),
                    "pCmdline": cmdline.replace('\x00', ' ').strip() if cmdline else "",
                    "pNumThreads": int(stat_parts[17]),
                    "tpNumThreads": t_stat_data,
                    "pCpuTimeUserMode": int(stat_parts[11]) * BaseCollector.JIFFIES_PER_SECOND,
                    "tpCpuTimeUserMode": t_stat_data,
                    "pCpuTimeKernelMode": int(stat_parts[12]) * BaseCollector.JIFFIES_PER_SECOND,
                    "tpCpuTimeKernelMode": t_stat_data,
                    "pChildrenUserMode": int(stat_parts[13]) * BaseCollector.JIFFIES_PER_SECOND,
                    "tpChildrenUserMode": t_stat_data,
                    "pChildrenKernelMode": int(stat_parts[14]) * BaseCollector.JIFFIES_PER_SECOND,
                    "tpChildrenKernelMode": t_stat_data,
                    "pVoluntaryContextSwitches": int(status_data.get('voluntary_ctxt_switches', 0)),
                    "tpVoluntaryContextSwitches": t_status_data,
                    "pNonvoluntaryContextSwitches": int(status_data.get('nonvoluntary_ctxt_switches', 0)),
                    "tpNonvoluntaryContextSwitches": t_status_data,
                    "pBlockIODelays": int(stat_parts[39]) * BaseCollector.JIFFIES_PER_SECOND,
                    "tpBlockIODelays": t_stat_data,
                    "pVirtualMemoryBytes": int(stat_parts[20]),
                    "tpVirtualMemoryBytes": t_stat_data,
                    "pResidentSetSize": rss_bytes,
                    "tpResidentSetSize": t_statm,
                })
            except Exception:
                BaseCollector.logger.exception("Failed to collect proc data")
                continue

        return {
            "pNumProcesses": process_count,
            "processes": [json.dumps(x).replace('"', '\"') for x in process_list]
        }

File: src/collectors/proc.go
package collectors

import (
	"path/filepath"
	"strconv"
	"strings"
	"syscall"
)

func CollectProcesses() (int64, []map[string]MetricValue) {
	dirs, err := filepath.Glob("/proc/[0-9]*")
	if err != nil {
		return 0, nil
	}

	pageSize := int64(syscall.Getpagesize())
	var processes []map[string]MetricValue

	for _, pidPath := range dirs {
		pid, err := strconv.ParseInt(filepath.Base(pidPath), 10, 64)
		if err != nil {
			continue
		}

		statData, tStat := ProbeFile(filepath.Join(pidPath, "stat"))
		rparenIndex := strings.LastIndex(statData, ")")
		if rparenIndex == -1 || len(statData) < rparenIndex+3 {
			continue
		}

		statParts := strings.Fields(statData[rparenIndex+2:])
		if len(statParts) < 40 {
			continue
		}

		cmdline, _ := ProbeFile(filepath.Join(pidPath, "cmdline"))
		status, tStatus := ParseProcKV(filepath.Join(pidPath, "status"), ":")
		statm, tStatm := ProbeFile(filepath.Join(pidPath, "statm"))

		var rssBytes int64
		if parts := strings.Fields(statm); len(parts) >= 2 {
			rssPages, _ := strconv.ParseInt(parts[1], 10, 64)
			rssBytes = rssPages * pageSize
		}

		pName := status["Name"]
		if pName == "" {
			if lp := strings.Index(statData, "("); lp != -1 && rparenIndex > lp {
				pName = statData[lp+1 : rparenIndex]
			}
		}

		processes = append(processes, map[string]MetricValue{
			"pId":                          {Value: pid, Time: tStat},
			"pName":                        {Value: pName, Time: tStat},
			"pCmdline":                     {Value: strings.ReplaceAll(cmdline, "\x00", " "), Time: tStat},
			"pNumThreads":                  NewMetricWithTime(parseInt64(statParts[17]), tStat),
			"pCpuTimeUserMode":             NewMetricWithTime(parseInt64(statParts[11])*JiffiesPerSecond, tStat),
			"pCpuTimeKernelMode":           NewMetricWithTime(parseInt64(statParts[12])*JiffiesPerSecond, tStat),
			"pChildrenUserMode":            NewMetricWithTime(parseInt64(statParts[13])*JiffiesPerSecond, tStat),
			"pChildrenKernelMode":          NewMetricWithTime(parseInt64(statParts[14])*JiffiesPerSecond, tStat),
			"pVoluntaryContextSwitches":    NewMetricWithTime(parseInt64(status["voluntary_ctxt_switches"]), tStatus),
			"pNonvoluntaryContextSwitches": NewMetricWithTime(parseInt64(status["nonvoluntary_ctxt_switches"]), tStatus),
			"pBlockIODelays":               NewMetricWithTime(parseInt64(statParts[39])*JiffiesPerSecond, tStat),
			"pVirtualMemoryBytes":          NewMetricWithTime(parseInt64(statParts[20]), tStat),
			"pResidentSetSize":             NewMetricWithTime(rssBytes, tStatm),
		})
	}

	return int64(len(processes)), processes
}
